docs/oppen_algorithm.md

```diff
@@ -0,0 +1,534 @@
++--------------+---------+-----------+----------+
+| size_buffer  |    ∞    |    -18    |    9     |
++--------------+---------+-----------+----------+
+
++-------------+-------+-------+
+| scan_stack  |   4   |   5   |
++-------------+-------+-------+
+
+Push BEGIN#2 to the print_stack.
+
++-------------+---------+---------+
+| print_stack | BEGIN#1 | BEGIN#2 |
++-------------+---------+---------+
+Output:
+------
+Hello$
+World!!!!!$
+------
+
++--------------+-----------+----------+
+| tokens_queue | LINEBREAK | STRING#3 |
++--------------+-----------+----------+
+| size_buffer  |    -18    |    9     |
++--------------+-----------+----------+
+
++-------------+-------+-------+
+| scan_stack  |   4   |   5   |
++-------------+-------+-------+
+
+Update the size_buffer to flush the line.
+
++--------------+-----------+----------+
+| tokens_queue | LINEBREAK | STRING#3 |
++--------------+-----------+----------+
+| size_buffer  |     ∞     |    9     |
++--------------+-----------+----------+
+
++-------------+-------+-------+
+| scan_stack  |   4   |   5   |
++-------------+-------+-------+
+
+Push LINEBREAK to the print_stack.
+
++-------------+---------+---------+
+| print_stack | BEGIN#1 | BEGIN#2 |
++-------------+---------+---------+
+Output:
+------
+Hello$
+World!!!!!$
+$
+------
+
++--------------+----------+
+| tokens_queue | STRING#3 |
++--------------+----------+
+| size_buffer  |    9     |
++--------------+----------+
+
++-------------+
+| scan_stack  |
++-------------+
+
+Push STRING#3 to the print stack.
+
++-------------+---------+---------+
+| print_stack | BEGIN#1 | BEGIN#2 |
++-------------+---------+---------+
+Output:
+------
+Hello$
+World!!!!!$
+            GoodBye$
+------
+
++--------------+
+| tokens_queue |
++--------------+
+| size_buffer  |
++--------------+
+
++-------------+
+| scan_stack  |
++-------------+
+```
+
+Since `scan_stack` is empty, all the remaining tokens will directly be forwarded to the `print_stack` without having to pass by the other data structures.
+
+```
+Push END#1    to the print_stack.
+
++-------------+---------+
+| print_stack | BEGIN#1 |
++-------------+---------+
+Output:
+------
+Hello$
+World!!!!!$
+            GoodBye$
+------
+
+Push END#2    to the print_stack.
+
++-------------+
+| print_stack |
++-------------+
+Output:
+------
+Hello$
+World!!!!!$
+            GoodBye$
+------
+```
+
+## Our extensions
+
+### Eager printing
+
+Oppen's original work and `ruby/PretttPrint`'s output differed when chaining groups.
+We extended Oppen's algorithm to allow groups to be printed eagerly: we will try to flush the current line every time an `END` token is encountered.
+
+Example for a maximum width of 13:
+```
+[
+  BEGIN,
+    BEGIN,
+      STRING = 'abc', BREAK, STRING = 'def',
+    END,
+
+    BEGIN,
+      STRING = 'ghi', BREAK, STRING = 'jkl',
+    END,
+  END,
+]
+
+without eager printing =>
+abc
+defghi jkl
+
+with eager printing =>
+abc defghi
+jkl
+```
+
```

Comment:
Take out of the code block, and write them normally:


Without eager printing:
```
...
```

With eager printing:
```
...
```